R4R: Updated release process documentation (#4133)
cosmos/cosmos-sdk: docs/RELEASE_PROCESS.md

### `cosmos/cosmos-sdk` Release Process

Unless otherwise stated, all of the following tasks will be undertaken by the release manager/coordinator:

- [ ] 1. Decide on release designation (are we doing a patch, or minor version bump)
- [ ] 2. Ensure that all commits/PRs which are destined for this release are merged to the `master` branch

> - [ ] 3. Create the release candidate branch (going forward known as **RC**) and ensure it's protected against pushing from anyone except the release manager/coordinator. **no PRs targeting this branch should be merged unless exceptional circumstances arise**

Reviewer: ```suggestion
- [ ] 3. Create the release candidate branch `rcN` (please start with `N=1`) (going forward known as **RC**) and ensure it's protected against pushing from anyone except the release manager/coordinator. **no PRs targeting this branch should be merged unless exceptional circumstances arise**
```

- [ ] 4. On the `RC` branch, use `clog` to prepare the `CHANGELOG.md`.
- [ ] 5. Summarize breaking API changes section under “Breaking Changes” section to the `CHANGELOG.md` to bring attention to any breaking API changes that affect RPC consumers.
- [ ] 6. Kick off a large round of simulation testing (e.g. 400 seeds for 2k blocks) 

> - [ ] 7. If errors are found in either #6 or #7 go back to #2 (*NOTE*: be sure to increment the `rcN`)

Reviewer: > (*NOTE*: be sure to increment the `rcN`)

Would this still be required?

Reviewer 2: Should be since further fixes or resolutions needs to be made.

Reviewer: There is no mention of `rcN` above, i.e. what is it that should be incremented? A lightweight tag? If so, why should we use it at all? This happens on a separate **RC** branch anyway.

Reviewer 2: It is (kind of) in step (3). No, nothing to do with tags. You create an entirely new branch off of the previous RC branch.

e.g.

1. Create a `release/v0.35.0-rc1` branch off of `master`
2. Run simulation; it fails and there is a bug (ughh ohhh)
3. Create a `release/v0.35.0-rc2` branch off of `release/v0.35.0-rc1`
4. Keep repeating this process until everything looks OK and then you can finally create `release/v0.35.0`

Reviewer: > You create an entirely new branch off of the previous RC branch.

I see, that makes sense then.

### `cosmos/cosmos-sdk` Release Process

> Unless otherwise stated, all of the following tasks will be undertaken by the release manager/coordinator

Reviewer 2: ```suggestion
Unless otherwise stated, all of the following tasks will be undertaken by the release manager/coordinator:
```

- [ ] 1. Decide on release designation (are we doing a patch, or minor version bump)
- [ ] 2. Ensure that all commits/PRs which are destined for this release are merged to the `master` branch

> - [ ] 3. Create the release candidate branch (going forward known as **RC**) and ensure it's protected against pushing from anyone except the release manager/coordinator. **no PRs targeting this branch should be merged unless exceptional circumstances arise**

Reviewer: **Hard Freeze** - correct, to speed up release

> - [ ] 4. On the `RC` branch, merge items in `PENDING.md` into the `CHANGELOG.md`. While doing this, make sure that each entry contains links to issues/PRs for each item

Reviewer: We use our homegrown tool for this: `clog` https://github.com/cosmos/cosmos-sdk/blob/develop/contrib/devtools/clog/main.go - I can elaborate further on its usage

Author: Should this line then become: "Use `clog` to prepare the `CHANGELOG.md`"?

Reviewer: `clog generate && rm .pending/` is the right command to finalize the changelog - and it should be the very last thing we do once QA passed and we are 120% sure we're going to make a new release.

Reviewer: Actually, when we fork `RC` out we should also `rm -rf .pending/` in `master`

- [ ] 5. Summarize breaking API changes section under “Breaking Changes” section to the `CHANGELOG.md` to bring attention to any breaking API changes that affect RPC consumers.

> - [ ] 6. Kick off 1 day of automated fuzz testing

Reviewer: fuzz testing = simulations. And it does not necessarily have to be one full day.

Reviewer 3: I think the tool @mircea-c wrote will allow us to do it much quicker!

```suggestion
- [ ] 6. Kick off a large round of simulation testing (e.g. 400 seeds for 2k blocks) 
```

> - [ ] 7. Assign 2 people to perform [buddy testing script](/docs/RELEASE_TEST_SCRIPT.md) and update the relevant documentation

Reviewer: I don't even remember the last time we've done this. This can be removed
